### ISSUE TITLE
Throw error when parsing the method with default parameters in v2.5.0

### ISSUE THREAD

**Issue by Sagilio** [davideicardi/DynamicExpresso#144]
Hello, I have written a little test, it will pass in `v2.4.1` and failed in `v2.5.0`. But I can not find any breaking change in the release log, 
Is this a bug or am I missing something?  Thank you!

### Test case:
```csharp
private delegate bool GFunction(string arg = null);

[Fact]
public void TestParseGFunction()
{
    static bool GetGFunction(string arg = null)
    {
        return arg is not null;
    };

    var interpreter = new Interpreter();
    interpreter.SetFunction("GFunction", (GFunction) GetGFunction);
    interpreter.SetVariable("arg", "arg");

    Assert.True((bool) interpreter.Eval("GFunction(arg)"));
    Assert.False((bool) interpreter.Eval("GFunction()"));
}
```

### Failed information:
```csharp
DynamicExpresso.Exceptions.ParseException
Argument list incompatible with delegate expression (at index 0).
   at DynamicExpresso.Parsing.Parser.ParseMethodGroupInvocation(MethodGroupExpression methodGroup, Int32 errorPos)
   at DynamicExpresso.Parsing.Parser.ParseUnary()
   at DynamicExpresso.Parsing.Parser.ParseMultiplicative()
   at DynamicExpresso.Parsing.Parser.ParseAdditive()
   at DynamicExpresso.Parsing.Parser.ParseTypeTesting()
   at DynamicExpresso.Parsing.Parser.ParseComparison()
   at DynamicExpresso.Parsing.Parser.ParseLogicalAnd()
   at DynamicExpresso.Parsing.Parser.ParseLogicalXor()
   at DynamicExpresso.Parsing.Parser.ParseLogicalOr()
   at DynamicExpresso.Parsing.Parser.ParseConditionalAnd()
   at DynamicExpresso.Parsing.Parser.ParseConditionalOr()
   at DynamicExpresso.Parsing.Parser.ParseConditional()
   at DynamicExpresso.Parsing.Parser.ParseAssignment()
   at DynamicExpresso.Parsing.Parser.ParseExpressionSegment()
   at DynamicExpresso.Parsing.Parser.ParseExpressionSegment(Type returnType)
   at DynamicExpresso.Parsing.Parser.Parse()
   at DynamicExpresso.Interpreter.ParseAsLambda(String expressionText, Type expressionType, Parameter[] parameters)
   at DynamicExpresso.Interpreter.Parse(String expressionText, Type expressionType, Parameter[] parameters)
   at DynamicExpresso.Interpreter.Eval(String expressionText, Type expressionType, Parameter[] parameters)
   at DynamicExpresso.Interpreter.Eval(String expressionText, Parameter[] parameters)
   at NetCasbin.UnitTest.UtilityTest.TestParseGFunction() in C:\Users\Sagilio\source\repos\OnWork\Casbin\Casbin.NET\NetCasbin.UnitTest\UtilityTest.cs:line 26
```

**Comment by davideicardi** [comment]
Yes, sorry. I suspect it is a regression due to #138 .
@metoule what do you think?

**Comment by metoule** [comment]
Yes, that seems like a side effect of my changes. I'll have a look at it!

**Comment by metoule** [comment]
Interestingly, I can only reproduce the issue if the static function is defined inside the `TestParseGFunction()` test method: if it's defined outside, then the test works properly (see below).

```csharp
private delegate bool GFunction(string arg = null);

static bool GetGFunction(string arg = null)
{
    return arg is not null;
};

[Fact]
public void TestParseGFunction()
{
    var interpreter = new Interpreter();
    interpreter.SetFunction("GFunction", (GFunction) GetGFunction);
    interpreter.SetVariable("arg", "arg");

    Assert.True((bool) interpreter.Eval("GFunction(arg)"));
    Assert.False((bool) interpreter.Eval("GFunction()"));
}
```

Note that DynamicExpresso is using C# 7.3, and defining a static function inside another one is only available for C# 8.0 or higher, so I'm not sure how I'll be able to add a unit test.

**Comment by metoule** [comment]
Here are my findings. Based on them, I'm not sure how to fix this issue. 

When invocating a delegate, there are two `MethodInfo` we can consider. The first is `Delegate.Method`: which you can get by simply by calling `((GFunction) GetGFunction).Method`. The second is to find the `Invoke` method of the delegate's type:

```csharp
```

**Comment by metoule** [comment]
Here are my findings. Based on them, I'm not sure how to fix this issue. 

When invocating a delegate, there are two `MethodInfo` we can consider. The first is `Delegate.Method`: which you can get by simply by calling `((GFunction) GetGFunction).Method`. The second is to find the `Invoke` method of the delegate's type:

```csharp
var flags = BindingFlags.Public | BindingFlags.DeclaredOnly | BindingFlags.Instance;
var invokeMethod = (MethodInfo)(((GFunction) GetGFunction).GetType().FindMembers(MemberTypes.Method, flags, Type.FilterName, "Invoke")[0]);
```

The problem is that they're not equivalent. 

When using the  `Delegate.Method`, the default value is missing, while the `Invoke` method is correct:

```csharp
private delegate bool GFunction(string arg = null);

static bool GetGFunction1(string arg = null)
{
	return arg != null;
}

[Test]
public void Set_function_With_Default_Params()
{
	static bool GetGFunction2(string arg = null)
	{
		return arg != null;
	}

	GFunction gFunc1 = GetGFunction1;
	GFunction gFunc2 = GetGFunction2;

	Assert.True(gFunc1.Method.GetParameters()[0].HasDefaultValue);
	Assert.False(gFunc2.Method.GetParameters()[0].HasDefaultValue); // should be true!
	
	var flags = BindingFlags.Public | BindingFlags.DeclaredOnly | BindingFlags.Instance;
	var invokeMethod1 = (MethodInfo)(gFunc1.GetType().FindMembers(MemberTypes.Method, flags, Type.FilterName, "Invoke")[0]);
	var invokeMethod2 = (MethodInfo)(gFunc1.GetType().FindMembers(MemberTypes.Method, flags, Type.FilterName, "Invoke")[0]);

	Assert.True(invokeMethod1.GetParameters()[0].HasDefaultValue);
	Assert.True(invokeMethod2.GetParameters()[0].HasDefaultValue);
}
```

On the other hand, the `Delegate.Method` preserve the params array attribute, while the `Invoke` method does NOT!

```csharp


[Test]
public void Set_function_With_Object_Params()
{
	var target = new Interpreter();

	// import static method with params array 
	var methodInfo = typeof(VariablesTest).GetMethod("Sum", BindingFlags.Static | BindingFlags.NonPublic);
	var types = methodInfo.GetParameters().Select(p => p.ParameterType).Concat(new[] { methodInfo.ReturnType });
	var del = methodInfo.CreateDelegate(Expression.GetDelegateType(types.ToArray()));
	target.SetFunction(methodInfo.Name, del);

	Assert.IsNotNull(del.Method.GetParameters()[0].GetCustomAttribute<ParamArrayAttribute>());

	var flags = BindingFlags.Public | BindingFlags.DeclaredOnly | BindingFlags.Instance;
	var invokeMethod = (MethodInfo)(del.GetType().FindMembers(MemberTypes.Method, flags, Type.FilterName, "Invoke")[0]);
	Assert.IsNull(invokeMethod.GetParameters()[0].GetCustomAttribute<ParamArrayAttribute>()); // should be not null!

	// the imported Sum function can be called with any parameters
	Assert.AreEqual(6, target.Eval<int>("Sum(1, 2, 3)"));
}

internal static int Sum(params int[] integers)
{
	return integers.Sum();
}
```

So I'm not sure how we can support for scenarii at once.

**Comment by davideicardi** [comment]
Thank you @metoule !

@Sagilio can you confirm that this bug is present only when defining function inside another one?

If this is a case I think that it shold not be a big problem ... maybe we can eval to list that as a limitation or if possible detect this situation and do not allow to register these kind of functions.

**Comment by metoule** [comment]
Another option would be to try both methods, I'll see if that works!

**Comment by Sagilio** [comment]
@davideicardi @metoule Thanks for analyzing!
Yes, it may only not work when defining a function inside another one.
But why it can be work properly in old versions?

**Comment by metoule** [comment]
@Sagilio version 2.5.0 includes a fix for #82 to allow the registration of multiple delegates with the same name, to simulate function overloads. To do that, I changed the way delegates are resolved when set via `SetFunction`. I introduced a regression, for which I apologize: I didn't realize that the `Delegate. Method` property and the `Invoke` method of the delegate type were different.

I fixed the issue via #145.
